nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -616,7 +675,187 @@ impl<'a> BlueprintBuilder<'a> {
             !removals.contains(&PhysicalDiskUuid::from_untyped_uuid(config.id))
         });
 
-        Ok(EnsureMultiple::Changed { added, removed })
+        Ok(EnsureMultiple::Changed { added, updated: 0, expunged: 0, removed })
+    }
+
+    /// Ensures that a sled in the blueprint has all the datasets it should.
+    ///
+    /// We perform the following process to decide what datasets should exist
+    /// in the blueprint during the planning phase:
+    ///
+    /// INPUT                    | OUTPUT
+    /// ----------------------------------------------------------------------
+    /// zpools in the blueprint  | blueprint datasets for debug, root filesystem
+    ///                          | (All zpools should have these datasets)
+    /// ----------------------------------------------------------------------
+    /// zones in the blueprint   | blueprint datasets for filesystems, durable data
+    ///                          | (These datasets are needed for zones)
+    /// ----------------------------------------------------------------------
+    /// discretionary datasets   | blueprint datasets for discretionary datasets
+    /// NOTE: These don't exist, |
+    /// at the moment            |
+    /// ----------------------------------------------------------------------
+    ///
+    /// From this process, we should be able to construct "all datasets that
+    /// should exist in the new blueprint".
+    ///
+    /// - If new datasets are proposed, they are added to the blueprint.
+    /// - If datasets are changed, they are updated in the blueprint.
+    /// - If datasets are not proposed, but they exist in the parent blueprint,
+    /// they are removed.
+    pub fn sled_ensure_datasets(
+        &mut self,
+        sled_id: SledUuid,
+        resources: &SledResources,
+    ) -> Result<EnsureMultiple, Error> {
+        let (mut additions, mut updates, expunges, removals) = {
+            let mut datasets_builder = BlueprintSledDatasetsBuilder::new(
+                self.log.clone(),
+                sled_id,
+                &self.datasets,
+                resources,
+            );
+
+            // Ensure each zpool has a "Debug" and "Zone Root" dataset.
+            let bp_zpools = self
+                .disks
+                .current_sled_disks(sled_id)
```

Comment:
At the moment, expunged disks are removed from the blueprint -- so this is dependent on the order of blueprint creation, but if  `sled_ensure_disks` is called first (it is) then this will only return non-expunged disks.

Reply:
I know we changed this PR to no longer remove expunged datasets. Are we still removing expunged disks? (Wondering if this question should be reopened after the removal changes.)